linera-service/src/linera-exporter/runloops/validator_exporter.rs

```diff
@@ -0,0 +1,215 @@
+
+            res = export_task.run(task_receiver) => res?,
+
+        };
+
+        Ok(())
+    }
+}
+
+impl<'a, S> ExportTask<'a, S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    fn new(
+        node: GrpcClient,
+        destination_id: DestinationId,
+        storage: &'a ExporterStorage<S>,
+    ) -> ExportTask<'a, S> {
+        ExportTask {
+            node,
+            destination_id,
+            storage,
+        }
+    }
+
+    async fn run(
+        &self,
+        mut receiver: Receiver<Arc<ConfirmedBlockCertificate>>,
+    ) -> anyhow::Result<()> {
+        let delivery = CrossChainMessageDelivery::NonBlocking;
+        while let Some(block) = receiver.recv().await {
+            let hash = block.hash();
+            let method = |certificate, delivery| {
+                self.node
+                    .handle_confirmed_certificate(certificate, delivery)
+            };
+
+            // As the linera-rpc client already contains a retry loop, the loop below is mainly for the blobs.
+            loop {
+                match dispatch!(method, log = hash, (*block).clone(), delivery) {
+                    Ok(_) => break,
+
+                    Err(NodeError::BlobsNotFound(blobs)) => self.upload_blobs(blobs).await?,
+
+                    Err(e) => Err(e)?,
+                }
+            }
+
+            self.storage.increment_destination(self.destination_id);
+        }
+
+        Ok(())
+    }
+
+    async fn upload_blobs(&self, blobs: Vec<BlobId>) -> anyhow::Result<()> {
+        let tasks = blobs.iter().map(|id| async {
+            match self.storage.get_blob(*id).await {
+                Err(e) => Err(e),
+                Ok(blob) => self
+                    .node
+                    .upload_blob((*blob).clone().into())
+                    .await
+                    .map_err(|e| ExporterError::GenericError(e.into()))
+                    .map(|_| ()),
+            }
+        });
+
+        let _ = try_join_all(tasks).await?;
+
+        Ok(())
+    }
+}
+
+struct TaskQueue<'a, S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    queue_size: usize,
+    start_height: usize,
+    storage: &'a ExporterStorage<S>,
+    buffer: Sender<Arc<ConfirmedBlockCertificate>>,
+}
+
+impl<'a, S> TaskQueue<'a, S>
+where
+    S: Storage + Clone + Send + Sync + 'static,
+{
+    fn new(
+        queue_size: usize,
+        destination_id: DestinationId,
+        storage: &'a ExporterStorage<S>,
+    ) -> (TaskQueue<'a, S>, Receiver<Arc<ConfirmedBlockCertificate>>) {
+        let start_height = storage.load_destination_state(destination_id) as usize;
+        let (sender, receiver) = tokio::sync::mpsc::channel(queue_size);
+
+        let queue = Self {
+            queue_size,
+            start_height,
+            storage,
+            buffer: sender,
+        };
+
+        (queue, receiver)
+    }
+
+    async fn run(&mut self) -> anyhow::Result<()> {
+        let mut index = self.start_height;
```

Comment:
```suggestion
        let mut last_processed_block_height = self.start_height;
```
is more clear, don't you think?

Reply:
index as in index for the logview which is more of a normal vector of BlobIds and blocks hashes from multiple chains in a sorted manner. we can't really use a block height here.

and it's not mutated by those futures (and FuturesOrdered are single threaded and the runtime too so it won't ever be sent to another thread)